In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
%cd '/content/gdrive/MyDrive/Colab Notebooks/BigDataAssignment4'

/content/gdrive/MyDrive/Colab Notebooks/BigDataAssignment4


In [ ]:
# import libraries
from sklearn import linear_model
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error
import numpy as np
import pandas as pd

In [ ]:
# read traning data
pd_train = pd.read_csv("exercise_3_training.csv")

# print the dataframe
print(pd_train)

    Unnamed: 0         x         y
0            0  0.000000  0.442580
1            1  0.052632  0.750379
2            2  0.105263  0.982555
3            3  0.157895  0.853643
4            4  0.210526  1.141193
5            5  0.263158  0.831381
6            6  0.315789  1.147227
7            7  0.368421  1.010907
8            8  0.421053  0.664553
9            9  0.473684  0.334169
10          10  0.526316 -0.058656
11          11  0.578947 -0.589828
12          12  0.631579 -0.569363
13          13  0.684211 -0.869054
14          14  0.736842 -1.166119
15          15  0.789474 -1.304733
16          16  0.842105 -0.830502
17          17  0.894737 -0.538124
18          18  0.947368 -0.485276
19          19  1.000000 -0.167024


In [ ]:
# create the input and output attributes
inp_attr = pd_train['x'].values.reshape(-1, 1)

# generate the nonlinear features
poly_order = 20
poly = PolynomialFeatures(poly_order,include_bias=False)
inp_feat = poly.fit_transform(inp_attr)
#print(inp_feat)
out_attr = pd_train['y'].values

In [ ]:
inp_feat.shape

(20, 20)

In [ ]:
# L2 penalty - vary alpha and observe the magnitude of the coeff and the errors
# Note that regularization coefficient - alpha=lambda (in the lecture)
# alpha = 0.0
# alpha = 0.1
# alpha = 100

l2 = Ridge(alpha=0.1)
l2.fit(inp_feat, out_attr)
print('coef = ', l2.coef_)

coef =  [-0.59024585 -1.67949412 -1.22932661 -0.57986637 -0.06644834  0.26968903
  0.45988645  0.54465164  0.55777382  0.52421925  0.46154418  0.38179648
  0.29311957  0.20094521  0.10883834  0.01908377 -0.06690923 -0.14832819
 -0.22476509 -0.29608326]


In [ ]:
# read testing data
pd_test = pd.read_csv("exercise_3_training.csv")

# create the input and output attributes
test_inp_attr = pd_test['x'].values.reshape(-1, 1)

# generate the nonlinear features
test_inp_feat = poly.fit_transform(test_inp_attr)
test_out_attr = pd_test['y'].values

In [ ]:
# L2 - model performance
l2_train_pred = l2.predict(inp_feat)
print('L2 - MSE train = ', mean_squared_error(l2_train_pred,out_attr))

l2_test_pred = l2.predict(test_inp_feat)
print('L2 - MSE test = ', mean_squared_error(l2_test_pred,test_out_attr))

L2 - MSE train =  0.0990607512903737
L2 - MSE test =  0.0990607512903737


## **1.Replace the line l2=Ridge(alpha=0.1) with l2=Lasso(alpha=0.1). Execute the notebook for alpha=0.1, and alpha=100.**

In [ ]:
# L2 penalty - vary alpha and observe the magnitude of the coeff and the errors
# Note that regularization coefficient - alpha=lambda (in the lecture)
# alpha = 0.0
# alpha = 0.1
# alpha = 100

l2=Lasso(alpha=0.1) # replacing l2=Lasso(alpha=0.1)
l2.fit(inp_feat, out_attr)
print('model parameters/coefficients = ', l2.coef_)

model parameters/coefficients =  [-1.03352367 -0.         -0.         -0.         -0.         -0.
 -0.         -0.         -0.         -0.         -0.         -0.
  0.          0.          0.          0.          0.          0.
  0.          0.        ]


In [ ]:
# read testing data
pd_test = pd.read_csv("exercise_3_training.csv")

# create the input and output attributes
test_inp_attr = pd_test['x'].values.reshape(-1, 1)

# generate the nonlinear features
test_inp_feat = poly.fit_transform(test_inp_attr)
test_out_attr = pd_test['y'].values

In [ ]:
# L2 - model performance
l2_train_pred = l2.predict(inp_feat)
print('L2 - MSE train = ', mean_squared_error(l2_train_pred,out_attr))

l2_test_pred = l2.predict(test_inp_feat)
print('L2 - MSE test = ', mean_squared_error(l2_test_pred,test_out_attr))

L2 - MSE train =  0.34234948898867806
L2 - MSE test =  0.34234948898867806


## **alpha=100.**

In [ ]:
# L2 penalty - vary alpha and observe the magnitude of the coeff and the errors
# Note that regularization coefficient - alpha=lambda (in the lecture)
# alpha = 0.0
#alpha = 0.1
#alpha = 100

l2=Lasso(alpha=100) # replacing l2=Lasso(alpha=100)
l2.fit(inp_feat, out_attr)
print('model parameters/coefficients = ', l2.coef_)

model parameters/coefficients =  [-0. -0. -0. -0. -0. -0. -0. -0. -0. -0. -0. -0. -0. -0. -0. -0. -0. -0.
 -0. -0.]


In [ ]:
# read testing data
pd_test = pd.read_csv("exercise_3_training.csv")

# create the input and output attributes
test_inp_attr = pd_test['x'].values.reshape(-1, 1)

# generate the nonlinear features
test_inp_feat = poly.fit_transform(test_inp_attr)
test_out_attr = pd_test['y'].values

In [ ]:
# L2 - model performance
l2_train_pred = l2.predict(inp_feat)
print('L2 - MSE train = ', mean_squared_error(l2_train_pred,out_attr))

l2_test_pred = l2.predict(test_inp_feat)
print('L2 - MSE test = ', mean_squared_error(l2_test_pred,test_out_attr))

L2 - MSE train =  0.647438408969314
L2 - MSE test =  0.647438408969314


**2. Add a new cell at the end of the notebook and add your answers to the following questions as comments. **

**2.1. What happens with the model parameters/coefficients when alpha=0.1?**

**Answer:**

1.    Lasso regression [ Least Absolute Shrinkage and Selection Operator],in cost
function it adds the penalty term. This term is calculated by absolute sum of the coefficients


2.     After applying L2=Ridge(alpha=0.1). The value of  model parameters/coefficients is: [-0.59024585 -1.67949412 -1.22932661 -0.57986637 -0.06644834  0.26968903 0.45988645  0.54465164  0.55777382  0.52421925  0.46154418  0.38179648 0.29311957  0.20094521  0.10883834  0.01908377 -0.06690923 -0.14832819 -0.22476509 -0.29608326]


3.     Also after applying the l2=Lasso(alpha=0.1) The value of  model parameters/coefficients is:   [-1.03352367 -0.         -0.         -0.         -0.         -0.  -0.         -0.         -0.         -0.         -0.         -0.   0.          0.          0.          0.          0.          0.  0.          0.   ]


4.     We clearly observe that after applying Lasso(alpha=0.1) value of model parameters/coefficients are changing the 1st parameter value is changing from [-0.59024585 ] to [-1.03352367], whereas remaining parameter value are changes from some value to the 0 (zero).
**2.2. What happens with the model parameters when alpha=100? Explain why is this happening.**

**Answer:**
1.  You can see that, as we increase the value of alpha, the magnitude of the coefficients decreases, where the coefficient value reaches zero.
2.   When alpha value is high there is a bigger penalty and for that reason the magnitude of coefficients are reduced.
3.    So, when alpha =100, the lasso regression  gives the null model in which all coefficient estimates equal zero
**2.3. How does MSE compares between alpha=0.1 and alpha=100? Explain which model you prefer.**

**Answer:**

1.   For our data when alpha = 0.1, then the lasso regression  gives the least squares fit, and when alpha becomes very large i.e alpha=100, the lasso gives the null model in which all coefficient estimates equal zero.

2.   MSE when alpha=0.1

     L2 - MSE train =  0.34234948898867806
     
     L2 - MSE test =  0.34234948898867806L2

3.   MSE when alpha=100

     L2 - MSE train =  0.647438408969314

     L2 - MSE test =  0.647438408969314

4.   We need the alpha value that minimizes the MSE because it will help to overcome overfitting.

5.    In this case the alpha=0.1 value gives the minimum value of Mean Square Error for training as well as test data. So, we select the alpha=0.1 to avoid overfitting.


## **3. Now replace the same line with l2=LinearRegression(). Execute the entire notebook.**

In [ ]:
# L2 penalty - vary alpha and observe the magnitude of the coeff and the errors
# Note that regularization coefficient - alpha=lambda (in the lecture)
# alpha = 0.0
#alpha = 0.1
#alpha = 100

l2=LinearRegression() # replace with l2=LinearRegression()
l2.fit(inp_feat, out_attr)
print('model parameters/coefficients = ', l2.coef_)

model parameters/coefficients =  [-2.60429857e+04  1.66021982e+06 -4.55463186e+07  7.25884218e+08
 -7.60880085e+09  5.60762719e+10 -3.02274990e+11  1.21971777e+12
 -3.72536198e+12  8.60617788e+12 -1.47811533e+13  1.79179180e+13
 -1.28403376e+13 -3.02767157e+11  1.35419924e+13 -1.81116496e+13
  1.34368586e+13 -6.14438987e+12  1.62920191e+12 -1.93081499e+11]


In [ ]:
# read testing data
pd_test = pd.read_csv("exercise_3_training.csv")

# create the input and output attributes
test_inp_attr = pd_test['x'].values.reshape(-1, 1)

# generate the nonlinear features
test_inp_feat = poly.fit_transform(test_inp_attr)
test_out_attr = pd_test['y'].values

In [ ]:
# L2 - model performance
l2_train_pred = l2.predict(inp_feat)
print('L2 - MSE train = ', mean_squared_error(l2_train_pred,out_attr))

l2_test_pred = l2.predict(test_inp_feat)
print('L2 - MSE test = ', mean_squared_error(l2_test_pred,test_out_attr))

L2 - MSE train =  8.037539242127623e-07
L2 - MSE test =  8.037539242127623e-07


**4. Add another new cell at the end of the notebook and add your answers to the following questions as comments.**

**4.1. What can you say about the magnitude of the model parameters?**

**Answer:**

1.   In the Linear Regression model,The magnitude of the  model parameters coefficient  shows the strength of the association.

2.  Also the sign i.e +ve and -ve indicate the direction of association
The coefficient value  1.79179180e+13 indicated the positive association between two variables, whereas a coefficient value -2.60429857e+04 indicated a weak negative association.

3. In addition to that , the coefficient value close to zero suggests no linear association between two variables.

4. Following are model parameters/coefficients =  [-2.60429857e+04  1.66021982e+06 -4.55463186e+07  7.25884218e+08
 -7.60880085e+09  5.60762719e+10 -3.02274990e+11  1.21971777e+12
 -3.72536198e+12  8.60617788e+12 -1.47811533e+13  1.79179180e+13
 -1.28403376e+13 -3.02767157e+11  1.35419924e+13 -1.81116496e+13
  1.34368586e+13 -6.14438987e+12  1.62920191e+12 -1.93081499e+11]

**4.2. What conclusion can you draw from the training and the testing MSE?**

**Answer:**
1.  L2 - MSE train =  8.037539242127623e-07
    
    L2 - MSE test =  8.037539242127623e-07


2. Training and testing  MSE gives the same values for the given data using Linear Regression Model

3.  The value is low which helps to eliminate the overfitting of the model.


**4.3. Is the linear regression model better than the previous two lasso models? From all the three models which one would you prefer and why?**

**Answer:**

1. Yes,  the linear regression model is better than the previous two lasso models.


2. From all three models Linear Regression model gives the lowest value of mean square error so, it will help to minimize the overfitting. For that reason I like to select the Linear Regression Model.
